common/BasicMultiQueueApplication.hpp

```diff
@@ -121,11 +121,30 @@ class BasicMultiQueueApplication : public virtual MonoDeviceApplication
 
-			// First thing to make sure we have is a compute queue (so nothing else fails allocation) which should be able to do image transfers of any granularity (transfer only queue families can have problems with that)
-			queue_req_t computeQueueRequirement = {.requiredFlags=queue_flags_t::EQF_COMPUTE_BIT,.disallowedFlags=queue_flags_t::EQF_NONE,.queueCount=1,.maxImageTransferGranularity={1,1,1}};
-			m_computeQueue.famIx = queueAllocator.allocateFamily(computeQueueRequirement,{queue_flags_t::EQF_GRAPHICS_BIT,queue_flags_t::EQF_TRANSFER_BIT,queue_flags_t::EQF_SPARSE_BINDING_BIT,queue_flags_t::EQF_PROTECTED_BIT});
-			// since we requested a device that has a compute capable queue family (unless `getQueueRequirements` got overriden) we're sure we'll get at least one family capable of compute
-			assert(m_computeQueue.famIx!=QueueAllocator::InvalidIndex);
+			// This is a sort-of allocator of queue indices for distinct queues
+			core::map<uint8_t, uint8_t> familyQueueCounts;
+
+			if (!isHeadlessCompute())
+			{
+				// TODO: Handle swapchain compatibility.
+				queue_req_t graphicsQueueRequirement = { .requiredFlags = queue_flags_t::EQF_GRAPHICS_BIT,.disallowedFlags = queue_flags_t::EQF_NONE,.queueCount = 1 };
+				// Place queue_flags_t::EQF_NONE in front in case there are no graphics queues without the unwanted flags.
+				m_graphicsQueue.famIx = queueAllocator.allocateFamily(graphicsQueueRequirement, { queue_flags_t::EQF_NONE, queue_flags_t::EQF_TRANSFER_BIT,queue_flags_t::EQF_SPARSE_BINDING_BIT, queue_flags_t::EQF_COMPUTE_BIT, queue_flags_t::EQF_PROTECTED_BIT });
+				assert(m_graphicsQueue.famIx != QueueAllocator::InvalidIndex);
+				m_graphicsQueue.qIx = familyQueueCounts[m_graphicsQueue.famIx]++;
+			}
+
+			// Make sure we have a compute queue (so nothing else fails allocation) which should be able to do image transfers of any granularity (transfer only queue families can have problems with that)
+			queue_req_t computeQueueRequirement = { .requiredFlags = queue_flags_t::EQF_COMPUTE_BIT,.disallowedFlags = queue_flags_t::EQF_NONE,.queueCount = 1,.maxImageTransferGranularity = {1,1,1} };
+			m_computeQueue.famIx = queueAllocator.allocateFamily(computeQueueRequirement, { queue_flags_t::EQF_GRAPHICS_BIT,queue_flags_t::EQF_TRANSFER_BIT,queue_flags_t::EQF_SPARSE_BINDING_BIT,queue_flags_t::EQF_PROTECTED_BIT });
+
+			if (m_computeQueue.famIx == QueueAllocator::InvalidIndex)
+			{
+				// Since we requested a device that has a compute capable queue family (unless `getQueueRequirements` got overriden) we're sure we'll get at least one family capable of compute.
+				// Going through this branch means we've selected that family for the graphics queue and no more queues can be allocated for this family. We have to alias.
+				m_logger->log("Not enough queue counts in families, had to alias the Compute Queue to Graphics!", system::ILogger::ELL_PERFORMANCE);
+				m_computeQueue = m_graphicsQueue;
+			}
```

Comment:
you need an else clause on both  the if-statements, because for non-headless you want to disallow the GRAPHICS queue capability when allocating the compute queue

Reply:
Actually best thing would be to allocate a fallback queue first that has following caps
```cpp
queue_req_t{ .requiredFlags = (isHeadlessCompute() ? EQF_GRAPHICS:EQF_NONE)|queue_flags_t::EQF_COMPUTE_BIT|queue_flags_t::EQF_transFER_BIT,.disallowedFlags = queue_flags_t::EQF_NONE,.queueCount = 1,.maxImageTransferGranularity = {1,1,1} };
```

then try-allocate separate transfer queues (with fallback of download to upload first), then try-allocate a separate compute queue without graphics capabilities, and finally a graphics queue without all the extra FORCED constraints

Reply:
> because for non-headless you want to disallow the GRAPHICS queue capability when allocating the compute queue, and you no longer need to require transfer granularities of 1,1,1

Yeah you're right about the transfer granularities but the current code already tries to find a compute queue without graphics in non-headless and only fallback to graphics capable queue if no such compute queue exists (for example having only combined graphics + compute and no async one). Or in such case the idea is (if no compute only queue exists) that we should just alias and not allocate another queue that also has graphics capabilities?

Reply:
>Actually best thing would be to allocate a fallback queue first that has following caps

Yeah that would simplify things. Pick one good for everything queue and then if we don't find async compute, DMA and graphics only queue for each alias to the fallback.